src/Futhark/Analysis/AccessPattern.hs

```diff
@@ -0,0 +1,607 @@
+{-# LANGUAGE LambdaCase #-}
+
+module Futhark.Analysis.AccessPattern
+  ( analyzeDimIdxPats,
+    analyzeFunction,
+    vnameFromSegOp,
+    Analyze,
+    IndexTable,
+    ArrayName,
+    DimIdxPat (DimIdxPat),
+    IndexExprName,
+    IterationType (Sequential, Parallel),
+    MemoryEntry (..),
+    BodyType (..),
+    SegOpName (SegmentedMap, SegmentedRed, SegmentedScan, SegmentedHist),
+  )
+where
+
+import Data.Either (partitionEithers)
+import Data.Foldable
+import Data.IntMap.Strict qualified as S
+import Data.Map.Strict qualified as M
+import Data.Maybe
+import Futhark.IR.Aliases
+import Futhark.IR.GPU
+import Futhark.IR.GPUMem
+import Futhark.IR.MC
+import Futhark.IR.MCMem
+import Futhark.IR.SOACS
+import Futhark.IR.Seq
+import Futhark.IR.SeqMem
+
+class Analyze rep where
+  analyzeOp :: Op rep -> (Context rep -> VName -> (Context rep, IndexTable rep))
+  orderIterationType :: IterationType rep -> IterationType rep -> Ordering
+
+-- | Map patterns of Segmented operations on arrays, to index expressions with
+-- their index descriptors.
+-- segmap(pattern) → A(pattern) → indexExpressionName(pattern) → [DimIdxPat]
+type IndexTable rep =
+  M.Map SegOpName (M.Map ArrayName (M.Map IndexExprName (MemoryEntry rep)))
+
+data SegOpName
+  = SegmentedMap (Int, VName)
+  | SegmentedRed (Int, VName)
+  | SegmentedScan (Int, VName)
+  | SegmentedHist (Int, VName)
+  deriving (Eq, Ord, Show)
+
+vnameFromSegOp :: SegOpName -> VName
+vnameFromSegOp (SegmentedMap (_, name)) = name
+vnameFromSegOp (SegmentedRed (_, name)) = name
+vnameFromSegOp (SegmentedScan (_, name)) = name
+vnameFromSegOp (SegmentedHist (_, name)) = name
+
+type ArrayName = VName
+
+type IndexExprName = VName
+
+-- | Each element in `dimensions` corresponds to an access to a given dimension
+-- in the given array, in the same order of dimensions.
+data MemoryEntry rep = MemoryEntry
+  { dimensions :: [DimIdxPat rep],
+    nest :: [BodyType]
+  }
+  deriving (Show)
+
+-- | Collect all features of access to a specific dimension of an array.
+newtype DimIdxPat rep = DimIdxPat
+  { -- | Set of VNames of gtid's that some access is variant to.
+    -- An empty set indicates that the access is invariant.
+    -- Tuple of patternName and nested `level` it is created at.
+    dependencies :: S.IntMap (VName, Int, IterationType rep)
+  }
+  deriving (Eq, Show)
+
+instance Semigroup (DimIdxPat rep) where
+  (<>) :: DimIdxPat rep -> DimIdxPat rep -> DimIdxPat rep
+  (<>) (DimIdxPat adeps) (DimIdxPat bdeps) =
```

Comment:
Infix: `DimIdxPat adeps <> DimIdxPat bdeps = ..,`

Reply:
Fixed in [99f26a27](https://github.com/diku-dk/futhark/pull/2027/commits/99f26a270718fd9e8bb3804a136054101a741116)